diff --git a/installer/chromeos-postinst b/installer/chromeos-postinst
index d093f1fda8..4ef5ceb209 100755
--- a/installer/chromeos-postinst
+++ b/installer/chromeos-postinst
@@ -26,6 +26,15 @@ cleanup() {
   done
 }
 
+get_disk_dev() {
+  /usr/bin/rootdev -d
+}
+
+get_part_num() {
+  local dev=$1
+  echo ${dev##*[a-z]}
+}
+
 main() {
   cd "${INSTALL_ROOT}" || exit 1
 
@@ -38,6 +47,9 @@ main() {
   done
 
   local install_dev="$1"
+  local root_dev=$(get_disk_dev)
+  local target_part_num=$(get_part_num $install_dev)
+  chroot . /usr/sbin/switch_root.sh $target_part_num $root_dev
   shift
   chroot . /usr/bin/cros_installer --type="postinst" \
          --install_dev="${install_dev}" --install_dir="/" "$@"
