Index: src/content/common/gpu_pre_sandbox_hook_linux.cc
===================================================================
--- src.orig/content/common/gpu_pre_sandbox_hook_linux.cc
+++ src/content/common/gpu_pre_sandbox_hook_linux.cc
@@ -437,7 +437,7 @@ void AddVulkanICDPermissions(std::vector
 
   static const char* const kReadOnlyICDList[] = {
       "intel_icd.x86_64.json", "nvidia_icd.json", "radeon_icd.x86_64.json",
-      "mali_icd.json", "freedreno_icd.aarch64.json"};
+      "mali_icd.json", "freedreno_icd.aarch64.json", "broadcom_icd.aarch64.json"};
 
   for (std::string prefix : kReadOnlyICDPrefixes) {
     permissions->push_back(BrokerFilePermission::ReadOnly(prefix));
@@ -637,6 +637,7 @@ void LoadVulkanLibraries() {
   dlopen("libvulkan_intel.so", dlopen_flag);
   dlopen("libGLX_nvidia.so.0", dlopen_flag);
   dlopen("libvulkan_freedreno.so", dlopen_flag);
+  dlopen("libvulkan_broadcom.so", dlopen_flag);
 }
 
 void LoadChromecastV4L2Libraries() {
