diff --git a/vc4.c b/vc4.c
old mode 100644
new mode 100755
index 71e73ea..f3030b3
--- a/vc4.c
+++ b/vc4.c
@@ -17,7 +17,10 @@
 #include "util.h"
 
 static const uint32_t render_target_formats[] = { DRM_FORMAT_ARGB8888, DRM_FORMAT_RGB565,
-						  DRM_FORMAT_XRGB8888 };
+						  DRM_FORMAT_XRGB8888, DRM_FORMAT_ABGR8888, DRM_FORMAT_XBGR8888,
+                          DRM_FORMAT_BGR565, DRM_FORMAT_ARGB1555, DRM_FORMAT_XRGB1555,
+                          DRM_FORMAT_YUV422, DRM_FORMAT_YVU422, DRM_FORMAT_YUV420,
+                          DRM_FORMAT_YVU420, DRM_FORMAT_NV12, DRM_FORMAT_NV16 };
 
 static int vc4_init(struct driver *drv)
 {
@@ -64,7 +67,7 @@ static void *vc4_bo_map(struct bo *bo, struct vma *vma, size_t plane, uint32_t m
 	struct drm_vc4_mmap_bo bo_map;
 
 	memset(&bo_map, 0, sizeof(bo_map));
-	bo_map.handle = bo->handles[0].u32;
+	bo_map.handle = bo->handles[plane].u32;
 
 	ret = drmCommandWriteRead(bo->drv->fd, DRM_VC4_MMAP_BO, &bo_map, sizeof(bo_map));
 	if (ret) {
