diff --git a/v8/src/parsing/parser-base.h b/v8/src/parsing/parser-base.h
index d30c3f1b1f..dc1d073186 100644
--- a/v8/src/parsing/parser-base.h
+++ b/v8/src/parsing/parser-base.h
@@ -715,7 +715,7 @@ class ParserBase {
   bool ContainsLabel(const ZonePtrList<const AstRawString>* labels,
                      const AstRawString* label) {
     DCHECK_NOT_NULL(label);
-    if (labels != nullptr) {
+    if (labels != nullptr && label != nullptr) {
       for (int i = labels->length(); i-- > 0;) {
         if (labels->at(i) == label) return true;
       }
