Index: rpi-linux-kernel/virt/kvm/Kconfig
===================================================================
--- rpi-linux-kernel.orig/virt/kvm/Kconfig
+++ rpi-linux-kernel/virt/kvm/Kconfig
@@ -47,7 +47,7 @@ config KVM_GENERIC_DIRTYLOG_READ_PROTECT
 
 config KVM_COMPAT
        def_bool y
-       depends on KVM && COMPAT && !(S390 || ARM64)
+       depends on KVM && COMPAT && !S390
 
 config HAVE_KVM_IRQ_BYPASS
        bool
