Index: kernel-rpi-6_1/drivers/gpu/drm/drm_auth.c
===================================================================
--- kernel-rpi-6_1.orig/drivers/gpu/drm/drm_auth.c
+++ kernel-rpi-6_1/drivers/gpu/drm/drm_auth.c
@@ -236,7 +236,8 @@ static int
 drm_master_check_perm(struct drm_device *dev, struct drm_file *file_priv)
 {
 	if (file_priv->was_master &&
-	    rcu_access_pointer(file_priv->pid) == task_tgid(current))
+	    (rcu_access_pointer(file_priv->pid) == task_tgid(current)
+        || drm_master_relax))
 		return 0;
 
 	if (!capable(CAP_SYS_ADMIN))
Index: kernel-rpi-6_1/drivers/gpu/drm/drm_drv.c
===================================================================
--- kernel-rpi-6_1.orig/drivers/gpu/drm/drm_drv.c
+++ kernel-rpi-6_1/drivers/gpu/drm/drm_drv.c
@@ -1058,6 +1058,8 @@ out:
 	return err;
 }
 
+bool drm_master_relax;
+
 static const struct file_operations drm_stub_fops = {
 	.owner = THIS_MODULE,
 	.open = drm_stub_open,
@@ -1090,6 +1092,9 @@ static int __init drm_core_init(void)
 
 	drm_debugfs_root = debugfs_create_dir("dri", NULL);
 
+  debugfs_create_bool("drm_master_relax", S_IRUSR | S_IWUSR,
+          drm_debugfs_root, &drm_master_relax);
+
 	ret = register_chrdev(DRM_MAJOR, "drm", &drm_stub_fops);
 	if (ret < 0)
 		goto error;
Index: kernel-rpi-6_1/include/drm/drm_drv.h
===================================================================
--- kernel-rpi-6_1.orig/include/drm/drm_drv.h
+++ kernel-rpi-6_1/include/drm/drm_drv.h
@@ -467,6 +467,8 @@ void *__devm_drm_dev_alloc(struct device
 			   const struct drm_driver *driver,
 			   size_t size, size_t offset);
 
+extern bool drm_master_relax;
+
 /**
  * devm_drm_dev_alloc - Resource managed allocation of a &drm_device instance
  * @parent: Parent device object
