Index: rpi-kernel/drivers/gpu/drm/vc4/vc4_drv.h
===================================================================
--- rpi-kernel.orig/drivers/gpu/drm/vc4/vc4_drv.h
+++ rpi-kernel/drivers/gpu/drm/vc4/vc4_drv.h
@@ -19,6 +19,7 @@
 
 #include "uapi/drm/vc4_drm.h"
 #define V3D_BIND (1<<7)
+#define CURSOR_TIMEOUT_MSEC 20
 #define VC4_KMS_OVERLAY_PLANE_NUM 8
 struct drm_device;
 struct drm_gem_object;
Index: rpi-kernel/drivers/gpu/drm/vc4/vc4_kms.c
===================================================================
--- rpi-kernel.orig/drivers/gpu/drm/vc4/vc4_kms.c
+++ rpi-kernel/drivers/gpu/drm/vc4/vc4_kms.c
@@ -302,6 +302,7 @@ static int vc4_atomic_commit(struct drm_
 {
 	struct vc4_dev *vc4 = to_vc4_dev(dev);
 	int ret;
+  unsigned long before_wait;
 
 	if (state->async_update) {
 		ret = down_interruptible(&vc4->async_modeset);
@@ -347,8 +348,9 @@ static int vc4_atomic_commit(struct drm_
 	}
 
 	if (!nonblock) {
+    before_wait = jiffies;
 		ret = drm_atomic_helper_wait_for_fences(dev, state, true);
-		if (ret) {
+		if (ret || (jiffies_to_msecs(jiffies - before_wait) > CURSOR_TIMEOUT_MSEC)) {
 			drm_atomic_helper_cleanup_planes(dev, state);
 			up(&vc4->async_modeset);
 			return ret;
