Index: kernel-raspberry/drivers/net/wireless/broadcom/brcm80211/brcmfmac/p2p.c
===================================================================
--- kernel-raspberry.orig/drivers/net/wireless/broadcom/brcm80211/brcmfmac/p2p.c
+++ kernel-raspberry/drivers/net/wireless/broadcom/brcm80211/brcmfmac/p2p.c
@@ -1466,7 +1466,8 @@ int brcmf_p2p_notify_action_frame_rx(str
 			       ETH_ALEN);
 	memcpy(mgmt_frame->sa, e->addr, ETH_ALEN);
 	mgmt_frame->frame_control = cpu_to_le16(IEEE80211_STYPE_ACTION);
-	memcpy(&mgmt_frame->u, frame, mgmt_frame_len);
+	unsafe_memcpy(&mgmt_frame->u, frame, mgmt_frame_len,
+    /* Bounds checked by mgmt_frame_size = offsetof u + mgmt_frame_len */);
 	mgmt_frame_len += offsetof(struct ieee80211_mgmt, u);
 
 	freq = ieee80211_channel_to_frequency(ch.control_ch_num,
