diff --git a/drivers/char/broadcom/vc_sm/vc_vchi_sm.c b/drivers/char/broadcom/vc_sm/vc_vchi_sm.c
index fda7e378652e..264e5678ed37 100644
--- a/drivers/char/broadcom/vc_sm/vc_vchi_sm.c
+++ b/drivers/char/broadcom/vc_sm/vc_vchi_sm.c
@@ -314,8 +314,8 @@ struct sm_instance *vc_vchi_sm_init(VCHI_INSTANCE_T vchi_instance)
 	set_user_nice(instance->io_thread, -10);
 	wake_up_process(instance->io_thread);
 
-	pr_debug("%s: success - instance 0x%x", __func__,
-		 (unsigned int)instance);
+	pr_debug("%s: success - instance 0x%p", __func__,
+		 instance);
 	return instance;
 
 err_close_services:
