--- a/src/android/camera_device.cpp
+++ b/src/android/camera_device.cpp
@@ -357,7 +357,9 @@
 	 * metadata.
 	 */
 	const auto &rotation = properties.get(properties::Rotation);
-	if (rotation) {
+  if (cameraConfigData && cameraConfigData->rotation != -1) {
+    orientation_ = cameraConfigData->rotation;
+  } else if (rotation) {
 		orientation_ = (360 - *rotation) % 360;
 		if (cameraConfigData && cameraConfigData->rotation != -1 &&
 		    orientation_ != cameraConfigData->rotation) {
@@ -738,7 +740,7 @@
 		for (const StreamConfiguration &cfg : *config)
 			LOG(HAL, Info) << " - " << cfg.toString();
 
-		return -EINVAL;
+		break;
 	case CameraConfiguration::Invalid:
 		LOG(HAL, Info) << "Camera configuration invalid";
 		return -EINVAL;
--- a/src/ipa/raspberrypi/meson.build
+++ b/src/ipa/raspberrypi/meson.build
@@ -2,8 +2,10 @@
 
 ipa_name = 'ipa_rpi'
 
+libboost=cc.find_library('boost', required :false)
 rpi_ipa_deps = [
     libcamera_private,
+    libboost,
     libatomic,
 ]
 
