Index: mesa-freedreno/src/gallium/drivers/v3d/v3dx_draw.c
===================================================================
--- mesa-freedreno.orig/src/gallium/drivers/v3d/v3dx_draw.c
+++ mesa-freedreno/src/gallium/drivers/v3d/v3dx_draw.c
@@ -1077,6 +1077,7 @@ v3d_draw_vbo(struct pipe_context *pctx,
                           V3D_DIRTY_COMPILED_GS_BIN |
                           V3D_DIRTY_COMPILED_GS |
                           V3D_DIRTY_COMPILED_FS |
+                          V3D_DIRTY_FRAMEBUFFER |
                           v3d->prog.cs->uniform_dirty_bits |
                           v3d->prog.vs->uniform_dirty_bits |
                           (v3d->prog.gs_bin ?
