Index: userland-e432bc3400401064e2d8affa5d1454aac2cf4a00/interface/mmal/core/CMakeLists.txt
===================================================================
--- userland-e432bc3400401064e2d8affa5d1454aac2cf4a00.orig/interface/mmal/core/CMakeLists.txt
+++ userland-e432bc3400401064e2d8affa5d1454aac2cf4a00/interface/mmal/core/CMakeLists.txt
@@ -11,7 +11,7 @@ add_library (mmal_core ${LIBRARY_TYPE}
    mmal_clock.c
 )
 
-target_link_libraries (mmal_core vcos)
+target_link_libraries (mmal_core vcos mmal_util)
 
 install(TARGETS mmal_core DESTINATION lib)
 install(FILES
