Index: adhd/cras/src/server/cras_alsa_io.c
===================================================================
--- a/adhd/cras/src/server/cras_alsa_io.c
+++ b/adhd/cras/src/server/cras_alsa_io.c
@@ -215,6 +215,11 @@ static const struct {
         .type = CRAS_NODE_TYPE_ALSA_LOOPBACK,
         .position = NODE_POSITION_INTERNAL,
     },
+    {
+      .name = "vc4-hdmi",
+      .type = CRAS_NODE_TYPE_HDMI,
+      .position = NODE_POSITION_EXTERNAL,
+    },
 };

 static inline int set_hwparams(struct cras_iodev* iodev) {
@@ -890,6 +895,7 @@ static void set_node_initial_state(struc

   node->volume = 100;
   node->type = CRAS_NODE_TYPE_UNKNOWN;
+  syslog(LOG_INFO, "test node:%s", node->name);
   // Go through the known names
   for (i = 0; i < ARRAY_SIZE(node_defaults); i++) {
     if (!strncmp(node->name, node_defaults[i].name,
@@ -2378,7 +2384,7 @@ int alsa_iodev_legacy_complete_init(stru
       !no_create_default_output_node(aio)) {
     if (first_internal_device(aio) && !has_node(aio, INTERNAL_SPEAKER) &&
         !has_node(aio, HDMI)) {
-      if (strstr(aio->common.base.info.name, HDMI)) {
+      if (strstr(aio->common.base.info.name, HDMI) || strstr(aio->common.base.info.name, "hdmi")) {
         new_output(aio, NULL, HDMI);
       } else {
         new_output(aio, NULL, INTERNAL_SPEAKER);
