diff --git a/init/upstart/cgroups.conf b/init/upstart/cgroups.conf
index d007d5e783..d0e576b838 100644
--- a/init/upstart/cgroups.conf
+++ b/init/upstart/cgroups.conf
@@ -105,6 +105,7 @@ script
   # Create a cpu cgroup for each VM type. Default cpu.shares is 1024. Limit the
   # VMs to 7/8ths of that initially.
   cgroup_dir="/sys/fs/cgroup/cpu/vms"
+  if [ -d "${cgroup_dir}" ]; then
   for vm_type in termina plugin arc ; do
     mkdir -p "${cgroup_dir}/${vm_type}"
     echo 896 > "${cgroup_dir}/${vm_type}/cpu.shares"
@@ -113,6 +114,7 @@ script
       chmod g+w "${cgroup_dir}/${vm_type}/${entry}"
     done
   done
+  fi
 
   schedtune_root_dir="/sys/fs/cgroup/schedtune"
   if [ -d "${schedtune_root_dir}" ]; then
