diff --git a/vm_tools/init/vm_concierge.conf b/vm_tools/init/vm_concierge.conf
index 8d151f259a..e7853176df 100644
--- a/vm_tools/init/vm_concierge.conf
+++ b/vm_tools/init/vm_concierge.conf
@@ -113,7 +113,7 @@ pre-start script
   mkdir -p /mnt/stateful_partition/unencrypted/apkcache
   chmod 700 /mnt/stateful_partition/unencrypted/apkcache
   chown 656360:656360 /mnt/stateful_partition/unencrypted/apkcache
-  chcon -t apkcache_file /mnt/stateful_partition/unencrypted/apkcache
+  #chcon -t apkcache_file /mnt/stateful_partition/unencrypted/apkcache
   mkdir -p /run/arcvm/apkcache
   nsenter --mount=/run/namespaces/mnt_concierge --no-fork \
     -- mount --bind /mnt/stateful_partition/unencrypted/apkcache \
